### ISSUE TITLE
Wayland: Window::set_mouse_visible doesn't work when window was clicked.

### ISSUE THREAD

**Issue by TobTobXX** [rust-windowing/winit#2273]
## Description
Calling [`Window::set_cursor_visible`](https://docs.rs/winit/latest/winit/window/struct.Window.html#method.set_cursor_visible) hides or shows the cursor. On Linux Wayland (GNOME 42), this doesn't work anymore though once you've clicked the window (at least) once.

## Demo code
I've composed a demo to reproduce it here: https://git.sr.ht/~tobtobxx/winit-wayland-mouse-demo

## Behavior
1. Compile & Run this code.
2. Correct behaviour: <kbd>C</kbd> hides and shows the cursor.
3. Click anywhare inside the window.
4. Incorrect behaviour: <kbd>C</kbd> doesn't work anymore.
5. Moving the mouse outside the window restores the correct behaviour.

(the hotkey <kbd>C</kbd> just toggles the call to `set_cursor_visible`: [source](https://git.sr.ht/~tobtobxx/winit-wayland-mouse-demo/tree/5811b89db33a76714046a77d87339fe9ab0c372e/item/src/main.rs#L39-46).)

## References
I first stumbled across this in veloren. It is reported there too: [Veloren, issue #1464](https://gitlab.com/veloren/veloren/-/issues/1464)

**Comment by kchibisov** [comment]
Do you have cursor theme installed and configured properly? I can't repro it with alacritty which hides the cursor on typing. Your demo works fine for me, C works regardless of clicking.

Consult https://wiki.archlinux.org/title/Cursor_themes#XDG_specification

**Comment by TobTobXX** [comment]
Yes, I had tried setting the cursor theme via the env vars `XCURSOR_THEME` and `XCURSOR_SIZE`. I have checked, I have the package `extra/adwaita-icon-theme` (Arch) installed.

As you suggested, I created the file `~/.icons/default/index.theme`:
```
[icon theme]
Inherits=Adwaita
```
That didn't work. I also edited the `~/.config/gtk-3.0/settings.ini` file to include the suggested line:
```
[Settings]
gtk-application-prefer-dark-theme=1
gtk-font-name=Noto Sans 11
gtk-icon-theme-name=Adwaita
gtk-theme-name=Adwaita
gtk-cursor-theme-name=Adwaita # <--- This is the new line
```
And I also checked the gnome-specific settings shown in the Arch Wiki:
```
$ gsettings get org.gnome.desktop.interface cursor-theme
'Adwaita'
$ gsettings get org.gnome.desktop.interface cursor-size
24
```

Note that showing and hiding the cursor works just fine, arbitrary many times. It breaks down once you clicked within the window and only starts working again once the cursor leaves the window.

**Comment by kchibisov** [comment]
I can't repro the issue no matter what I do with your example, neither with alacritty. What compositor are you using?

Also, you could try `strace` it and see what it loads for cursor theme.

**Comment by TobTobXX** [comment]
> neither with alacritty

Weird, I downloaded Alacritty and enabled mouse hiding only to observe the same broken behaviour: Works, until I clicked once within the Window. Which compositor are you using?

> What compositor are you using?

Not so sure which Arch package exactly is the gnome compositor, so here all that seem relevant:
```
gnome-desktop 1:42.1-1
gnome-desktop-4 1:42.1-1
gnome-desktop-common 1:42.1-1
gnome-session 42.0-1
gnome-shell 1:42.0-1
xdg-desktop-portal-gnome 42.1-1
```


> Also, you could try strace it and see what it loads for cursor theme.

Did it, the full log is attached. Grepped for `Adwaita` and found that it (sucessfully) read at least these files:
 - `/home/tobtobxx/.icons/default/index.theme` file: ASCII text
 - `/usr/share/icons/default/index.theme` file: ASCII text
 - `/usr/share/icons/Adwaita/index.theme` file: ASCII text, with very long lines (1372)
 - `/usr/share/icons/Adwaita/cursors/left_ptr` file: X11 cursor

```strace
[...]
statx(AT_FDCWD, "/home/tobtobxx/.icons/default", AT_STATX_SYNC_AS_STAT, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFDIR|0755, stx_size=4096, ...}) = 0
openat(AT_FDCWD, "/home/tobtobxx/.icons/default/index.theme", O_RDONLY|O_CLOEXEC) = 18
statx(18, "", AT_STATX_SYNC_AS_STAT|AT_EMPTY_PATH, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0644, stx_size=31, ...}) = 0
lseek(18, 0, SEEK_CUR)                  = 0
read(18, "[icon theme]\nInherits=Adwaita\n\n", 31) = 31
read(18, "", 32)                        = 0
close(18)                               = 0
[...]
statx(AT_FDCWD, "/usr/share/icons/default", AT_STATX_SYNC_AS_STAT, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFDIR|0755, stx_size=4096, ...}) = 0
openat(AT_FDCWD, "/usr/share/icons/default/index.theme", O_RDONLY|O_CLOEXEC) = 18
statx(18, "", AT_STATX_SYNC_AS_STAT|AT_EMPTY_PATH, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0644, stx_size=30, ...}) = 0
lseek(18, 0, SEEK_CUR)                  = 0
read(18, "[Icon Theme]\nInherits=Adwaita\n", 30) = 30
read(18, "", 32)                        = 0
close(18)                               = 0
[...]
statx(AT_FDCWD, "/usr/share/icons/Adwaita", AT_STATX_SYNC_AS_STAT, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFDIR|0755, stx_size=4096, ...}) = 0
openat(AT_FDCWD, "/usr/share/icons/Adwaita/index.theme", O_RDONLY|O_CLOEXEC) = 18
statx(18, "", AT_STATX_SYNC_AS_STAT|AT_EMPTY_PATH, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0644, stx_size=7425, ...}) = 0
lseek(18, 0, SEEK_CUR)                  = 0
read(18, "[Icon Theme]\nName=Adwaita\nCommen"..., 7425) = 7425
read(18, "", 32)                        = 0
close(18)                               = 0
[...]
statx(AT_FDCWD, "/usr/share/icons/Adwaita/cursors/left_ptr", AT_STATX_SYNC_AS_STAT, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0644, stx_size=69120, ...}) = 0
openat(AT_FDCWD, "/usr/share/icons/Adwaita/cursors/left_ptr", O_RDONLY|O_CLOEXEC) = 18
statx(18, "", AT_STATX_SYNC_AS_STAT|AT_EMPTY_PATH, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0644, stx_size=69120, ...}) = 0
lseek(18, 0, SEEK_CUR)                  = 0
read(18, "Xcur\20\0\0\0\0\0\1\0\5\0\0\0\2\0\375\377\30\0\0\0L\0\0\0\2\0\375\377"..., 69120) = 69120
read(18, "", 32)                        = 0
brk(0x564852bd4000)                     = 0x564852bd4000
lseek(16, 0, SEEK_END)                  = 1024
ftruncate(16, 3328)                     = 0
write(16, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 2304) = 2304
close(18)                               = 0
```

[strace.txt](https://github.com/rust-windowing/winit/files/8598066/strace.txt)

**Comment by TobTobXX** [comment]
To Debug, I added `dbg!(&cursor_icon)` line in the `platform_impl::wayland::window::shim::WindowHandle::set_cursor_visible()` function ([here](https://github.com/rust-windowing/winit/blob/ea09d1d10e59f7b44dedfd18a9170a01a4930768/src/platform_impl/linux/wayland/window/shim.rs#L342)).

I can't say whether this output is what is expected, but it doesn't seem odd:
```
[/home/tobtobxx/projects/winit/src/platform_impl/linux/wayland/window/shim.rs:342] &cursor_icon = Some(
    Default,
)
[/home/tobtobxx/projects/winit/src/platform_impl/linux/wayland/window/shim.rs:342] &cursor_icon = None
[/home/tobtobxx/projects/winit/src/platform_impl/linux/wayland/window/shim.rs:342] &cursor_icon = Some(
    Default,
)
[/home/tobtobxx/projects/winit/src/platform_impl/linux/wayland/window/shim.rs:342] &cursor_icon = None
[/home/tobtobxx/projects/winit/src/platform_impl/linux/wayland/window/shim.rs:342] &cursor_icon = Some(
    Default,
)
[/home/tobtobxx/projects/winit/src/platform_impl/linux/wayland/window/shim.rs:342] &cursor_icon = None
[/home/tobtobxx/projects/winit/src/platform_impl/linux/wayland/window/shim.rs:342] &cursor_icon = Some(
    Default,
)
```
(I triggered the bug in between these lines.)

**Comment by kchibisov** [comment]
>Weird, I downloaded Alacritty and enabled mouse hiding only to observe the same broken behaviour: Works, until I clicked once within the Window. Which compositor are you using?

I'm using sway. Could you repro the issue on it?

>Also btw, another behavior I noticed working correctly is when I set the size using the XCURSOR_SIZE=60 env var. This correctly enlarges the mouse. Issue with hiding still persists though.

Hm, that means that cursor theme is indeed loaded correctly. The output you've linked seems about right...

**Comment by TobTobXX** [comment]
Hah, maybe I gottem (maybe not, I'm by no means an expert of Wayland and all that stuff):

Read this last line of the docs:
[`wayland_client::protocol::wl_pointer::WlPointer::set_cursor()`](https://docs.rs/wayland-client/0.29.4/wayland_client/protocol/wl_pointer/struct.WlPointer.html):

> The serial parameter must match the latest wl_pointer.enter serial number sent to the client. Otherwise the request will be ignored.

I logged the serial number of each call and when (once) clicked, the serial number changes. If I read correctly, due to [this line](https://github.com/rust-windowing/winit/blob/ea09d1d10e59f7b44dedfd18a9170a01a4930768/src/platform_impl/linux/wayland/seat/pointer/handlers.rs#L152). I removed the line and it works now!

It seems like `set_cursor()` requires the serial number of the last **wl_pointer.enter** event, while (probably) other functions need the serial number from any latest event. Thus I think removing this line will probably break stuff, but this is something that you probably know better.

This also doesn't definitively explain the discrepancy between sway and gnome. Could be just a laxer implementation of sway, but that'd need to be verified.
